Smallrye context propagation

FroMage opened the pull request:
Fixes #467:

- Added `quarkus-smallrye-context-propagation` module
- Support for context propagation for ArC, RESTEasy, Transaction
- Static-init-time initialisation of Smallrye-CP
- Transactions are committed/rolled-back at the end of async requests if any
- New abstraction for async request listeners in ArC
- Injection of `ManagedExecutor` and `ThreadContext` for `ALL` contexts, using the Quarkus thread pool.

stuartwdouglas committed:
Request context needs to last for the life of the request

FroMage committed:
Added Smallrye-CP module

FroMage committed:
Fixed Arc test for Quarkus and MP-CP

manovotn committed:
Implement CDI context propagation for Arc.

manovotn committed:
Correction to tests, add test case for cleared CDI context.

FroMage committed:
Handle async requests in transaction interceptor

FroMage committed:
Test async transactions

FroMage committed:
ArcContextProvider: use constant

FroMage committed:
UndertowDeploymentTemplate: rework Stuart's patch

FroMage committed:
Support injected ThreadContext and ManagedExecutor backed by quarkus thread pool

FroMage committed:
Formatting

FroMage committed:
Initialise SmallRye CP at static init time

FroMage committed:
Test CP natively in main integration tests

FroMage committed:
Use abstraction for async request in ArC rather than depend on undertow/resteasy

FroMage committed:
I think we should not listen for async requests if we're throwing

FroMage commented:
We can merge this before we support the configurable CDI injection since it depends on #2243.

FroMage committed:
Make sure we follow the MP-CP spec and disallow our container-managed executor to be shut down

Since it's the underlying executor service which must be shut down, by Quarkus

FroMage commented:
Brilliant, the tests don't pass on CI, even though they pass locally :(

FroMage commented:
It _could_ be related to the `CompletionStage` abstraction I just made, because we might be invoked in the wrong thread while before I was careful to be notified in the servlet async listener which undertow executed in the proper thread. But for those particular failing tests, there's no async request, so I've no idea why this would even differ from the prior sync behaviour :(

manovotn commented on extensions/smallrye-context-propagation/runtime/src/main/java/io/quarkus/smallrye/context/runtime/SmallRyeContextPropagationProvider.java:
```diff
@@ -0,0 +1,47 @@
+
+    @Produces
+    @Singleton
+    public ThreadContext getAllThreadContext() {
```
I'd rather not have these as they will instantly clash with addition of injection support in https://github.com/quarkusio/quarkus/issues/2243. We should pick one of the other approaches discussed on issue - either SR-specific API (but that kills the spec purpose somewhat) or make it so that users can only spawn executors from Quarkus thread pool (which would allow users to stick to spec API only). I am not sure if it even makes sense to let users create executors from both sources?
Also, with current design, you differentiate the pools used (Quarkus vs new) by using builder or injection. That doesn't seem very user friendly to me.

FroMage replied:
Well no they won't because we don't have #2243 yet and when we do I can remove those. As far as users are concerned their code will keep working the same.

manovotn commented on extensions/smallrye-context-propagation/runtime/src/main/java/io/quarkus/smallrye/context/runtime/SmallRyeContextPropagationProvider.java:
```diff
@@ -0,0 +1,47 @@
+            }
+
+            @Override
+            public List<Runnable> shutdownNow() {
```
This behaviour is not expected by TCKs, see for instance https://github.com/eclipse/microprofile-context-propagation/blob/master/tck/src/main/java/org/eclipse/microprofile/context/tck/ManagedExecutorTest.java#L1737

FroMage replied:
Here's the javadoc that says I must do this:

```java
 * <p>Managed executors that are created with the {@link Builder} or created for
 * injection into applications via CDI must support the life cycle methods, including:
 * awaitTermination, isShutdown, isTerminated, shutdown, shutdownNow.
 * The application must invoke <code>shutdown</code> or <code>shutdownNow</code> to terminate
 * the life cycle of each managed executor that it creates, once the managed executor is
 * no longer needed. When the application stops, the container invokes <code>shutdownNow</code>
 * for any remaining managed executors that the application did not already shut down.
 * The shutdown methods signal the managed executor that it does not need to remain active to
 * service subsequent requests and allow the container to properly clean up associated resources.</p>
 *
 * <p>Managed executors which have a life cycle that is scoped to the container,
 * including those obtained via mechanisms defined by EE Concurrency, must raise
 * IllegalStateException upon invocation of the aforementioned life cycle methods,
 * in order to preserve compatibility with that specification.</p>
```

FroMage replied:
Defined on `ManagedExecutor`.

manovotn replied:
That's not really what the TCK assumes though.
Also the bit you refer to is mostly for the sake of EE Conc compatibility which we don't really care about here. David's executor view allows to recognize that state of that ([here](https://github.com/jbossas/jboss-threads/blob/master/src/main/java/org/jboss/threads/ViewExecutor.java#L43-L46)), why not just forward the calls there?

FroMage replied:
Well, then the TCK either does not test container-managed executors or it's invalid. That part of the spec is pretty important.

Now, I didn't know the view did not foward the shutdown to the underlying executor. In the case of container-managed executors that's fine (although it's contra spec which says it should throw), but I made it so SR-CP would use the view on any user-provided executor, and I'm not sure it's the right semantics here, because if the user creates a thread pool executor himself and we make a view out of it, the user may expect that shutting down our `ManagedExecutor` would also shut down the thread pool he created, but it does not. Do you think it's an issue @dmlloyd ?

manovotn replied:
Hmm, that does sounds like an issue for those created through builders.
With beans, I suppose we could just weave in a disposer method that will perform this on bean destruction (which, with application scoped beans, is on application shutdown).is definitely

Or, we would have to make sure the threads in the pool are destroyed when not used (basically spawned on demand, dropped when idle for some time).

manovotn replied:
> Or, we would have to make sure the threads in the pool are destroyed when not used (basically spawned on demand, dropped when idle for some time).

Uhm, hold on, that's actually how I did it back there. We should already have that in place here - https://github.com/smallrye/smallrye-context-propagation/blob/master/core/src/main/java/io/smallrye/context/SmallRyeManagedExecutor.java#L43-L49

FroMage replied:
Filed https://github.com/eclipse/microprofile-context-propagation/issues/158 to clarify the spec/TCK mismatch, and possibly question the spec's decision about throwing.

FroMage replied:
Well, the spec says:

    When the application stops, the container invokes <code>shutdownNow</code>
    for any remaining managed executors that the application did not already shut down.

So it looks like we need to keep hold of them all.

> Uhm, hold on, that's actually how I did it back there. We should already have that in place here

Yes but that's only for those that are created by the builder. Not for the views we create for user-specified executors (with the new SM-CP API to specify an executor in the builder).

manovotn committed:
Introduce Arc.isRunning as a quick way to determine CDI availibility. Use that in ArcContextProvider.

manovotn commented on extensions/smallrye-context-propagation/runtime/src/main/java/io/quarkus/smallrye/context/runtime/context/provider/ArcContextProvider.java:
```diff
@@ -0,0 +1,129 @@
+     */
+    private boolean isCdiAvailable() {
+        try {
+            return CDI.current() != null;
```
After discussing this with Martin, we decides its ineffective and we should instead introduce and use `ArcContainer.isRunning()`.

manovotn replied:
@FroMage Apparently I am not allowed to contribute to your existing branch or I am doing it wrong. Here is the commit you need to cherry-pick here - https://github.com/manovotn/quarkus/commit/7a54d8476e7cb4a571c3245914e257919f528161

FroMage replied:
Just added you as colab, you should be able to push to this branch now.

manovotn replied:
Ok, will try

manovotn replied:
Pushed the commit here

FroMage commented on extensions/smallrye-context-propagation/runtime/src/main/java/io/quarkus/smallrye/context/runtime/SmallRyeContextPropagationProvider.java:
```diff
@@ -0,0 +1,47 @@
+@ApplicationScoped
+public class SmallRyeContextPropagationProvider {
+
+    private volatile ManagedExecutorImpl managedExecutor;
```
I copied this from the `PgPoolProvider` class.

FroMage replied:
I think it can be re-set in dev mode?

manovotn replied:
How would you re-set it except hard reflection?
Looks superfluous to me.

FroMage replied:
It probably is, I can ask @tsegismont why he put it in `PgPoolProvider`?.

FroMage replied:
I've removed it.

FroMage commented:
I'm more worried about why CI fails :(

@manovotn could you try to do a build of this branch and run the tests? They all pass for me.

manovotn commented:
@FroMage tried locally and it passes.

FroMage commented:
> @FroMage tried locally and it passes.

Thanks. It passed CI, but not for Windows, but it does not seem related. I'll re-run Windows.

FroMage commented:
Well, windows build failed, this time for a real multi-threading issue in our transaction test. And we can't reproduce it locally, probably because our machines are too fast. DAMN.

FroMage committed:
Use latest SR-CP

dmlloyd reviewed (approved):
Seems basically OK

FroMage committed:
Fix version of SM-CP-api

FroMage committed:
Use new RESTEasy CP module

FroMage committed:
Latest SR-CP changes

FroMage committed:
Added support for extensions too

FroMage commented:
Rebased on master, updated to latest resteasy context propagation provider, added support for extensions.

FroMage committed:
Make sure we suspend/resume the transaction when going async

FroMage commented:
OK I found what I had forgotten to make async transactions work: suspend it after going async, and resume it when notified of async completion. Hopefully that should make CI pass.

dmlloyd commented:
Maybe you want to squash in the "fucking formatter" changes? :)

FroMage committed:
Removed unneeded volatile

FroMage committed:
Make sure we set the default executor to the quarkus one

dmlloyd commented:
The JDK11 failure seems related.

FroMage commented:
So I think this PR is ready for merging, pending docs that I have to write.

FroMage commented:
> The JDK11 failure seems related.

Uh, related to what?

FroMage commented:
Damnit. I swear the tests were running last time I pushed :(

FroMage commented:
@manovotn @mkouba after I rebased on master, this test failed: https://github.com/quarkusio/quarkus/pull/2321/files#diff-49fe536ebeac25a592ac959dc502952dR39

It's something related to behaviour changed in ArC I think, because this is testing availability of a request scope bean in a request that is not yet async, so I really think I did not cause that change (I hope). Do you have any idea what happened?

mkouba commented:
@FroMage I believe the `RequestBean` got removed because it's not injected anywhere etc. The simplest solution is to `@Inject RequestBean` in the `ContextPropagationResource`. I see no real value in using `Arc.container().instance()` here. In fact, the bean must always be available inside the resource method. And since a proxy is injected for `@RequestScoped` beans you should be able to use the injected reference inside the `thenApplyAsync` callback too. In other words, there is no need to use `Arc.container().instance()` at all (if the request context is propagated correcly).

FroMage commented:
>  I believe the RequestBean got removed because it's not injected anywhere

Hum, but this test was working until very recently. That's very confusing, why did it not get removed before?

> And since a proxy is injected for @RequestScoped beans you should be able to use the injected reference inside the thenApplyAsync callback too. In other words, there is no need to use Arc.container().instance() at all (if the request context is propagated correcly).

But then am I correctly testing context propagation? In orther words: is the proxy using the ArC thread-local underneath?

mkouba commented:
> Hum, but this test was working until very recently. That's very confusing, why did it not get removed before?

Hm, the test is part of the `main` integration test suite which also includes MP Fault Tolerance extension which contained `@Inject Instance<Object>` until recently (and this basically eliminates the bean removal optimization). Note that `ContextEndpoint` does inject the bean.

> But then am I correctly testing context propagation? In orther words: is the proxy using the ArC thread-local underneath?

Yes. Client proxies always get the "current context" (context associated with the current thread), then get the "current instance" and delegate all method calls to this instance.

FroMage commented:
> Hm, the test is part of the main integration test suite which also includes MP Fault Tolerance extension which contained @Inject Instance<Object> until recently (and this basically eliminates the bean removal optimization). Note that ContextEndpoint does inject the bean.

OK thanks, that explains the mystery :)

> Yes. Client proxies always get the "current context" (context associated with the current thread), then get the "current instance" and delegate all method calls to this instance.

OK, but at what point? When injected, or whenever a method is called on the proxy? I can't inject inside the async callback, so if I inject it in the main thread and it's already resolved to the current instance when I call the method in the async callback I'm not testing anything.

FroMage committed:
Make sure the request bean is not removed

mkouba commented:
> OK, but at what point? When injected, or whenever a method is called on the proxy? I can't inject inside the async callback, so if I inject it in the main thread and it's already resolved to the current instance when I call the method in the async callback I'm not testing anything.

An injected client proxy is not bound to a thread local. In fact, we do reuse a client proxy reference whenever a normal scoped bean is injected. The current context is looked up whenever you invoke a method. In other words, it is safe to inject a bean using the main thread and call the bean in your callback.

FroMage commented:
Oh damn, I'm writing the guide about this and I notice the transaction guide tells people to use `@Transactional` on non-rest beans. So now I'm wondering about the request scope that applies to these method calls.

For example, suppose this code:

```
@Path
public class Rest {

 @Inject
 Service service;

 @Transactional
 @GET
 public CompletionStage<String> hello(){
   service.call();
   return ...;
 }
}

@ApplicationScoped
public class Service {
  @Transactional
  public void call() {
    // ...
  }
}
```

IMO it's clear that the transaction started for `hello()` should only end when the async result is complete, when the async request is finished. But `service.call()` sits in another transaction, and I'm pretty sure users expect that second transaction to be ended after the call to `service.call()` and not at the end of the request. Now, in the future it may be possible to declare that transactional services that return a `CompletionStage` might want thei transaction to last until that `CS` is resolved, but this would be another feature entirely.

Does anyone know if within the `service.call()` method we have a servlet request scope? And if yes, is it the same as the REST request ? I suppose I have to test this, but perhaps @mkouba knows?

mkouba commented:
Hm, I do not see any async operation here ;-). `service.call()` is invoked inside the resource method, on the same thread, so I would expect it to be invoked within the transaction started by the resource method - no new transaction is started (`TxType.REQUIRED` is default).

dmlloyd commented:
As an aside - I've approved this but we should also get an approval from at least one of the other requested reviewers.

mkouba commented:
I think that @manovotn is the right person to chase ;-)

FroMage committed:
Test that only intercepted request methods delay transaction commit

manovotn reviewed (approved):
Looks OK

gsmet commented:
@FroMage how risky is it to merge this one just before the release?

I was thinking that maybe we should merge it just after to have 2 weeks of hindsight but I haven't taken a look at the PR.

manovotn commented:
My 0.02$ are that changes here should revolve around context propagation in its various forms so it seems pretty safe for other existing use cases. That being said, the propagation itself needs some victi-, uh I mean early adopters to tell us if it works as expected. So from that perspective, the sooner it gets into some release, the better :)

gsmet commented:
@FroMage told me on Zulip this PR needs some further work so marking it with `coding-in-progress` for now. Please remove the tag once it's ready.

mkouba commented:
So I have finally some spare cycles to review this PR:
1. `ArcContainer#isCurrentRequestAsync()/getAsyncRequestNotifier()` seems to be only used in `TransactionalInterceptorBase`. Therefore, I see no reason to declare the methods on the `ArcContainer` at all. There should be probably a dedicated `@Singleton` bean (`AsyncRequesStatus` or something like that) in the `smallrye-context-propagation` runtime module so that any component could simply inject the bean or lookup it programatically and invoke its `isCurrentRequestAsync()/getAsyncRequestNotifier()` methods.
2. It seems there could be only one `AsyncRequestStatusProvider` but multiple `AsyncRequestNotifierProvider`s? Sounds a little bit weird. This would mean that there is only one type of request that could be potentially "turned" asynchronous but multiple providers that signal that an async request is completed. I would expect that there could be only one `AsyncRequestStatusProvider` per the current thread. But my knowledge is very limited in this area so it's probably OK ;-)

mkouba reviewed (dismissed):
Move the `ArcContainer#isCurrentRequestAsync()/getAsyncRequestNotifier()` methods into a dedicated bean.

soberich commented:
Is it possible to further "propagate" on this to reach Kotlin coroutines support?
I checked out locally `FroMage:smallrye-context-propagation` and tried with CompletionStage. it went  fine. But calling `withContext(CoroutineContext)`
```kotlin
Dispatchers.IO {
}
```
causes
```console
2019-05-17 08:47:06,519 ERROR [org.jbo.res.res.i18n] (vert.x-eventloop-thread-0) RESTEASY002025: Unknown exception while executing GET /fruits/employee: javax.enterprise.context.ContextNotActiveException: interface javax.enterprise.context.RequestScoped
	at io.quarkus.hibernate.orm.runtime.RequestScopedEntityManagerHolder_ClientProxy.delegate(Unknown Source)
	at io.quarkus.hibernate.orm.runtime.RequestScopedEntityManagerHolder_ClientProxy.getOrCreateEntityManager(Unknown Source)
	at io.quarkus.hibernate.orm.runtime.entitymanager.TransactionScopedEntityManager.getEntityManager(TransactionScopedEntityManager.java:91)
	at io.quarkus.hibernate.orm.runtime.entitymanager.TransactionScopedEntityManager.unwrap(TransactionScopedEntityManager.java:388)
	at io.quarkus.hibernate.orm.runtime.entitymanager.ForwardingEntityManager.unwrap(ForwardingEntityManager.java:238)
	at com.example.web.rest.jakarta.BlockingServiceImpl$listEmployees$2.invokeSuspend(BlockingServiceImpl.kt:46)
```

whether with transaction or not, whether in CDI bean or RestEasy resource.

FroMage commented:
@mkouba if I move things to their own bean in SR-CP then RESTEasy and Undertow need to depend on it.

FroMage commented:
@soberich I've no idea about this. It doesn't look like `CoroutineContext` is the functional interface you're looking for, though. You need to find the function that will invoke the coroutine, perhaps its' `Job`? Otherwise we may need to register contexts from MP-CP into `CoroutineContext`, but that would require some spec changes.

FroMage commented:
> Hm, I do not see any async operation here ;-). 

The request is async because it returns a `CompletionStage`.

> service.call() is invoked inside the resource method, on the same thread, so I would expect it to be invoked within the transaction started by the resource method - no new transaction is started (TxType.REQUIRED is default).

Right, I meant to write `REQUIRES_NEW`. I just tested it and my PR causes the new transaction to be delayed until the end of the request, so I need a way to figure out if the interceptor is calling a RESTEasy endpoint (`Rest.hello`) or not (`Service.call`).

mkouba commented:
@FroMage you could move `AsyncRequestNotifierProvider` and `AsyncRequestStatusProvider` to the core runtime module (it's imho more a core feature than a DI feature) and then select the beans in `AsyncRequesStatus` from the SM CP runtime.

FroMage commented:
OK. But the other question is much more important. I've no idea how to fix that.

FroMage commented:
I suppose I could use `ctx.getMethod()` from the `InvocationContext` and compare it with the equivalent reflection from RESTEasy but that looks very hackish.

mkouba commented:
> The request is async because it returns a CompletionStage.

@FroMage so RESTEasy automatically invokes the method on a "worker" thread?

> I need a way to figure out if the interceptor is calling a RESTEasy endpoint (Rest.hello) or not (Service.call)

Hm, that's tricky. I see no easy way to do this. Maybe use reflection to find an HTTP method annotation declared on a method (`@GET` etc.)? AFAIK every resource method must declare an HTTP method annotation.

FroMage commented:
> @FroMage so RESTEasy automatically invokes the method on a "worker" thread?

No, but it turns the request async: it suspends it. I agree the terminology here is confusing.

FroMage committed:
Only check if the intercepted method turned the request async

FroMage commented:
I found a solution by comparing the intercepted method with RESTEasy's invoked method, which means it won't work with Servlet, but that's fine.

FroMage committed:
Replaced implementation based on services with wrapping CS/Publisher

Must use SR-type-converters and SR-reactive-streams-operators to do that

FroMage commented:
OK, following our discussion on Zulip, I changed the implementation away from service providers in favour of intercepting the annotated method's return type:

- If it is a `CompletionStage` or `Publisher`: hook into it and end the transaction when it's done
- Otherwise, if it is not null, try to find a type converter (with SmallRye-Reactive-Type-Converters) to turn it into a `CompletionStage` or `Publisher`, hook similarly and turn back to original type before returning the value (yes that's lame)
- If it's not those types, nor something that can be converted to those types: old behaviour

mkouba commented:
I still think that `AsyncRequestNotifierProvider`/`AsyncRequestStatusProvider` should not be part of the the `independent-projects/arc` module. See https://github.com/quarkusio/quarkus/pull/2321#issuecomment-491800553. Maybe `core/runtime`?

FroMage committed:
Added CP build item

FroMage committed:
Moved ArcContextProvider to ArC

FroMage committed:
Added CP to extensions list

FroMage committed:
Added deps to respective CP providers

RESTEasy->context provider
Narayana->context provider
RS-operators->rxjava2 propagator and quarkus-SM-CP

FroMage committed:
Added documentation about context propagation

FroMage commented:
I've added docs.

The current import situation is as follows:

- `quarkus-arc` includes its context provider, depends only on mp-cp (not smallrye impl)
- `quarkus-resteasy` includes its context provider via `resteasy-context-propagation`, depends only on mp-cp (not smallrye impl)
- `quarkus-narayana` includes its context provider, depends only on mp-cp (not smallrye impl)
    - it also depends on reactive-type-converters and reactive-streams-operators to support pluggable async types
- `quarkus-smallrye-reactive-streams-operators` implies a dependency on `quarkus-smallrye-context-propagation` and `smallrye-context-propagation-propagators-rxjava2` for automatic context propagation

If you are not using `quarkus-smallrye-reactive-streams-operators` and want context propagation for a `CompletionStage` then you need to import `quarkus-smallrye-context-propagation` but that's all: all the providers will be included by the relevant dependency (as shown above).

I think that's a fair model.

FroMage commented:
> I still think that AsyncRequestNotifierProvider/AsyncRequestStatusProvider should not be part of the the independent-projects/arc module. See #2321 (comment). Maybe core/runtime?

I removed them entirely, no? Did I forget something?

FroMage commented:
I did forget to remove them, sorry :)

FroMage committed:
Remove unneeded async request providers

FroMage commented:
Sorry, fixed now!

soberich commented:
@FroMage 

I am sorry, I expressed myself not clear, the question was 
"are you guys going to support context propagation to kotlin coroutines?"

with 0.16.0
you got 
```console
2019-06-05 00:44:13,862 ERROR [org.jbo.res.res.i18n] (vert.x-eventloop-thread-0) RESTEASY002025: Unknown exception while executing GET /example/employee: javax.enterprise.context.ContextNotActiveException: interface javax.enterprise.context.RequestScoped
	at io.quarkus.hibernate.orm.runtime.RequestScopedEntityManagerHolder_ClientProxy.delegate(Unknown Source)
	at io.quarkus.hibernate.orm.runtime.RequestScopedEntityManagerHolder_ClientProxy.getOrCreateEntityManager(Unknown Source)
	at io.quarkus.hibernate.orm.runtime.entitymanager.TransactionScopedEntityManager.getEntityManager(TransactionScopedEntityManager.java:93)
	at io.quarkus.hibernate.orm.runtime.entitymanager.TransactionScopedEntityManager.contains(TransactionScopedEntityManager.java:279)
	at io.quarkus.hibernate.orm.runtime.entitymanager.ForwardingEntityManager.contains(ForwardingEntityManager.java:138)
	at io.quarkus.hibernate.orm.panache.runtime.JpaOperations.persist(JpaOperations.java:29)
	at io.quarkus.hibernate.orm.panache.runtime.JpaOperations.persist(JpaOperations.java:25)
	at io.quarkus.hibernate.orm.panache.PanacheRepositoryBase.persist(PanacheRepositoryBase.java:37)
```
with panache or when call any method on injected `EntityManager` with `PersistenceContext` annotation. E.g. when you shift execution to say 
```kotlin
Dispatchers.IO {
}
```
so, in pseudo code
this works
```kotlin
        em.clear() //or any other method
        continueOn(Dispatchers.IO)
        //other code
```
this does not
```kotlin
        continueOn(Dispatchers.IO)
        em.clear() //or any other method
        //other code
```

FroMage commented:
@soberich I don't know how kotlin coroutines work, but if you're willing to provide a patch to https://github.com/smallrye/smallrye-context-propagation I'll merge it. I think you should be able to implement it like the propagator plugin for RxJava2: https://github.com/smallrye/smallrye-context-propagation/tree/master/propagators-rxjava2
